File: packages/python/jobping/transport_layer.py
"""TransportLayer abstractions for JobPing metadata and semantic messages."""

from __future__ import annotations

from abc import ABC, abstractmethod
from typing import Any, Literal, NotRequired, TypedDict

from jobping.envelope import (
    EnvelopeType,
    JobPingEnvelope,
)


JOBPING_JOB_ID_HEADER = "x-jobping-job-id"


class TransportCarrier(TypedDict):
    headers: NotRequired[dict[str, str]]
    envelope: NotRequired[JobPingEnvelope]


MockCarrier = TransportCarrier


class TransportMessage(TypedDict):
    kind: str
    job_id: str
    data: NotRequired[Any]


class TransportLayer(ABC):
    @abstractmethod
    def attach_job_id(self, carrier: TransportCarrier | None, job_id: str) -> TransportCarrier:
        raise NotImplementedError

    @abstractmethod
    def extract_job_id(self, carrier: TransportCarrier | None) -> str | None:
        raise NotImplementedError

    @abstractmethod
    def attach_envelope(
        self,
        carrier: TransportCarrier | None,
        envelope: JobPingEnvelope,
    ) -> TransportCarrier:
        raise NotImplementedError

    @abstractmethod
    def extract_envelope(self, carrier: TransportCarrier | None) -> JobPingEnvelope | None:
        raise NotImplementedError

    @abstractmethod
    def send_envelope(self, envelope: JobPingEnvelope) -> None:
        raise NotImplementedError

    @abstractmethod
    async def recv_envelope(
        self,
        *,
        job_id: str | None = None,
        type: EnvelopeType | None = None,
        timeout: float = 1.0,
    ) -> JobPingEnvelope:
        raise NotImplementedError

    @abstractmethod
    def send_message(self, message: TransportMessage) -> None:
        raise NotImplementedError

    @abstractmethod
    async def recv_message(
        self,
        *,
        kind: str | None = None,
        job_id: str | None = None,
        timeout: float = 1.0,
    ) -> TransportMessage:
        raise NotImplementedError

File: packages/js/transport_layer.mjs
// TransportLayer defines how JobPing metadata and semantic messages move.
//
// Concrete transports may use HTTP headers, WebSocket messages, SSE+POST,
// Kafka, Redis, RabbitMQ, or another carrier. This layer does not manage
// JPItem lifecycle and does not inspect business results.

export const JOBPING_JOB_ID_HEADER = "x-jobping-job-id";

export class TransportLayer {
  constructor() {
    if (new.target === TransportLayer) {
      throw new Error("TransportLayer is abstract; use a concrete implementation");
    }
  }

  attachJobId() {
    throw new Error("TransportLayer.attachJobId() must be implemented");
  }

  extractJobId() {
    throw new Error("TransportLayer.extractJobId() must be implemented");
  }

  attachEnvelope() {
    throw new Error("TransportLayer.attachEnvelope() must be implemented");
  }

  extractEnvelope() {
    throw new Error("TransportLayer.extractEnvelope() must be implemented");
  }

  sendEnvelope() {
    throw new Error("TransportLayer.sendEnvelope() must be implemented");
  }

  recvEnvelope() {
    throw new Error("TransportLayer.recvEnvelope() must be implemented");
  }

  sendMessage() {
    throw new Error("TransportLayer.sendMessage() must be implemented");
  }

  recvMessage() {
    throw new Error("TransportLayer.recvMessage() must be implemented");
  }
}
